source/distributed/fully_distributed_tria.cc

```diff
@@ -0,0 +1,525 @@
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/memory_consumption.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+namespace GridGenerator
+{
+  template <int dim, int spacedim>
+  void
+  hyper_cube(Triangulation<dim, spacedim> &tria,
+             const double                  left,
+             const double                  right,
+             const bool                    colorize);
+} // namespace GridGenerator
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace internal
+    {
+      /**
+       * Check if `parent` is the parent of `child`.
+       */
+      template <int dim>
+      bool
+      is_parent(typename CellId::binary_type parent,
+                typename CellId::binary_type child)
+      {
+        if (child[0] != parent[0])
+          return false; // has same coarse cell
+        if ((child[1] >> 2) != ((parent[1] >> 2) + 1))
+          return false; // level has one difference
+
+
+        const unsigned int n_child_indices = parent[1] >> 2;
+        const unsigned int children_per_value =
+          sizeof(CellId::binary_type::value_type) * 8 / dim;
+        unsigned int child_level  = 0;
+        unsigned int binary_entry = 2;
+
+        // Loop until all child indices have been written
+        while (child_level < n_child_indices)
+          {
+            Assert(binary_entry < child.size(), ExcInternalError());
+
+            for (unsigned int j = 0; j < children_per_value; ++j)
+              {
+                if ((((child[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)) !=
+                    (((parent[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)))
+                  return false;
+
+
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+
+            ++binary_entry;
+          }
+        return true;
+      }
+    } // namespace internal
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::reinit(
+      const ConstructionData<dim, spacedim> &construction_data)
+    {
+      // clear internal data structures
+      this->coarse_cell_id_to_coarse_cell_index_vector.clear();
+      this->coarse_cell_index_to_coarse_cell_id_vector.clear();
+
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.coarse_cell_vertices.empty() == true)
```

Comment:
It's a bit hard to read because you ask whether a negative question ("empty") is true :-) I'd say that's one occasion where we could omit the explicit `== true`.

Separately, the comment above states the opposite of what the conditional asks ;-)